### ISSUE TITLE
Disconnection issue

### ISSUE THREAD

**Issue by alexandreroche** [pusher/pusher-websocket-swift#131]
Hi there,

I'm getting an issue where auto-reconnecting is not working. If I use my app for a bit, then background it (for a few minutes perhaps), and then return, I sometimes get this in my log (this is from logging within the changeConnectionState and debugLog delegate methods)....

```
Websocket is disconnected. Error: The operation couldn’t be completed. Socket is not connected
pusher connection changes connected to reconnecting
pusher debugLog: [PUSHER DEBUG] Network reachable so will setup reconnect attempt
pusher debugLog: [PUSHER DEBUG] Waiting 0.0 seconds before attempting to reconnect (attempt 1 of 6)
2017-03-27 19:47:26.099311 Vibes[6956:1522519] [] nw_socket_get_input_frames recvmsg(fd 4, 1024 bytes): [57] Socket is not connected
2017-03-27 19:47:26.099490 Vibes[6956:1522519] [] nw_socket_get_input_frames recvmsg(fd 11, 1024 bytes): [57] Socket is not connected
2017-03-27 19:47:26.099569 Vibes[6956:1522519] [] nw_socket_get_input_frames recvmsg(fd 19, 1024 bytes): [57] Socket is not connected
2017-03-27 19:47:26.099665 Vibes[6956:1522519] [] nw_socket_get_input_frames recvmsg(fd 22, 1024 bytes): [57] Socket is not connected
2017-03-27 19:47:26.099722 Vibes[6956:1522519] [] nw_socket_get_input_frames recvmsg(fd 4, 1024 bytes): [57] Socket is not connected
2017-03-27 19:47:26.115581 Vibes[6956:1522519] [] nw_endpoint_handler_add_write_request [26.1 2600:9000:201e:b200:15:8b9b:bf00:21.443 failed socket-flow (satisfied)] cannot accept write requests
2017-03-27 19:47:26.123091 Vibes[6956:1522522] [] __tcp_connection_write_eof_block_invoke Write close callback received error: [22] Invalid argument
2017-03-27 19:47:26.125074 Vibes[6956:1522519] [] nw_endpoint_handler_add_write_request [27.1 2400:cb00:2048:1::6810:d0a5.443 failed socket-flow (satisfied)] cannot accept write requests
2017-03-27 19:47:26.125315 Vibes[6956:1522522] [] __tcp_connection_write_eof_block_invoke Write close callback received error: [22] Invalid argument
2017-03-27 19:47:26.126280 Vibes[6956:1522519] [] nw_endpoint_handler_add_write_request [30.1 35.163.223.46:443 failed socket-flow (satisfied)] cannot accept write requests
2017-03-27 19:47:26.127003 Vibes[6956:1522520] [] __tcp_connection_write_eof_block_invoke Write close callback received error: [22] Invalid argument
```

From that point on, my connection is pwned and my app can no longer receive messages from Pusher. Any idea what could be causing this?

Thanks!
Alexandre

**Comment by matthewtsmith** [comment]
I'm seeing the same issue. My app gets stuck in the state of `reconnectingWhenNetworkBecomesReachable` when returning from the background. I believe I've seen it when my network changes as well. At that point I can't re-connect to Pusher. The user must force close the application.

**Comment by hamchapman** [comment]
First off, sorry for being slow to comment - I've been away for a couple of weeks.

Thanks to both of you for the information though. Do either of you have projects that can be shared (or just a cut down version of a project) that reproduces the error?

On that note, is the error constantly occurring when attempting to reconnect or is it sporadic?

I can try and recreate this locally to see if I can see what the issue is if neither of you are able to provide a project.

Let me know - thanks!

**Comment by matthewtsmith** [comment]
@hamchapman Unfortunately I do not have a project I can share. For me it seems sporadic, most of the time it seems to function as expected. Usually I come across it by accident while debugging. 

The last time I was able to reproduce it was on the simulator. I opened my laptop on a public wifi network that has a splash page. I accepted the terms on the splash page and my computer successfully connected to the network. All network functions in my app worked but Pusher was stuck in the `reconnectingWhenNetworkBecomesReachable` state. 
I then tried calling pusher.disconnect() and pusher.connect(). After calling pusher.connect() it still remained in that state.

I know that isn't a great repro case but its the most info I've been able to get because I had the debugger connected.

Our app is used a lot in higher ed/public sectoer so I would imagine that the users who are reporting this have similar conditions whether its needing VPN or Wifi Splash pages. 

Hopefully @alexandreroche can shine some more light on this. Unfortunately, once it happens the functionality reliant on pusher becomes unusable (which for us is a lot).

**Comment by matthewtsmith** [comment]
We are still having this issue occur frequently. After going through the source I see that once the reconnectMaxAttempts are hit ( which defaults to 6 ), the connection will no longer try to ever re-connect. 

https://github.com/pusher/pusher-websocket-swift/blob/6a542bd68a76a7d1c7620bb7b8dcce6eb0591c80/Source/PusherWebsocketDelegate.swift#L59

I believe this is the exact issue we are running into. For whatever reason their network conditions are such that it retries 6 times and then they must kill the app for the connection to try again. I don't see a good reason to ever stop retrying, even if at a minimum interval. In addition, even calling `channel.disconnect()` on the channel does not reset the reconnect counter.

**Comment by hamchapman** [comment]
See this part of the readme for information about controller reconnection behaviour: https://github.com/pusher/pusher-websocket-swift#reconnection

In short:

> `public var reconnectAttemptsMax: Int? = 6` - if you set this to nil then there is no maximum number of reconnect attempts and so attempts will continue to be made with an exponential backoff (based on number of attempts)

and

> `public var maxReconnectGapInSeconds: Double? = nil` - if you want to set a maximum length of time (in seconds) between reconnect attempts then set this property appropriately

**Comment by rudysuharyadi** [comment]
@hamchapman that two property are optional old data type and did not bridge to objective-C.

**Comment by hamchapman** [comment]
@rudysuharyadi I'm not sure I understand what you mean. Why would they not bridge to Objective-C?

**Comment by rudysuharyadi** [comment]
@hamchapman because objective-C cannot represent nil value for old data type, so they not bridge those property to objective-C. If you use pusher-websocket-swift library on objective-C, you dont have access to those property even though its public property already.

**Comment by hamchapman** [comment]
@rudysuharyadi thanks for the explanation - good to know! I've merged #163 which makes `OCReconnectAttemptsMax` and `OCMaxReconnectGapInSeconds` available on `PusherConnection` instances to let Objective-C users get and set the `reconnectAttemptsMax` and `maxReconnectGapInSeconds` properties.

I'll be releasing a new version later today that contains this addition.
